core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockStore.java

```diff
@@ -0,0 +1,380 @@
+        return false;
+      }
+    } finally {
+      mLock.unlock();
+    }
+    blockInfo.closeReaderOrWriter();
+    return blockInfo.getMeta().getCommitPending();
+  }
+
+  /**
+   * Releases the access token of this block by removing this (sessionId, blockId) pair from the
+   * store.
+   *
+   * @param sessionId the session ID
+   * @param blockId the block ID
+   */
+  public void releaseAccess(long sessionId, long blockId) {
+    mLock.lock();
+    try {
+      Key key = new Key(sessionId, blockId);
+      mBlocks.remove(key);
+      Set<Long> blockIds = mSessionIdToBlockIds.get(sessionId);
+      if (blockIds != null) {
+        blockIds.remove(blockId);
+      }
+      Set<Long> sessionIds = mBlockIdToSessionIds.get(blockId);
+      if (sessionIds != null) {
+        sessionIds.remove(sessionId);
+      }
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  /**
+   * Cleans up all the block information(e.g. block reader/writer) that belongs to this session.
+   *
+   * @param sessionId the session ID
+   */
+  public void cleanupSession(long sessionId) {
+    Set<Long> blockIds;
+    mLock.lock();
+    try {
+      blockIds = mSessionIdToBlockIds.get(sessionId);
+      if (blockIds == null) {
+        return;
+      }
+    } finally {
+      mLock.unlock();
+    }
+
+    for (Long blockId : blockIds) {
+      try {
+        // Note that we don't need to explicitly call abortBlock to cleanup the temp block
+        // in Alluxio block store because they will be cleanup by the session cleaner in the
+        // Alluxio block store.
+        cleanup(sessionId, blockId);
+        releaseAccess(sessionId, blockId);
+      } catch (Exception e) {
+        LOG.warn("Failed to cleanup UFS block {}, session {}.", blockId, sessionId);
+      }
+    }
+  }
+
+  /**
+   * Creates a block reader that reads from UFS and optionally caches the block to the Alluxio
+   * block store.
+   *
+   * @param sessionId the client session ID that requested this read
+   * @param blockId the ID of the block to read
+   * @param offset the read offset within the block (NOT the file)
+   * @param noCache if set, do not try to cache the block in the Alluxio worker
+   * @return the block reader instance
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the
+   * {@link UnderFileSystemBlockStore}
+   * @throws IOException if any I/O errors occur
+   */
+  public BlockReader getBlockReader(final long sessionId, long blockId, long offset,
+      boolean noCache) throws BlockDoesNotExistException, IOException {
+    final BlockInfo blockInfo;
+    mLock.lock();
+    try {
+      blockInfo = getBlockInfo(sessionId, blockId);
+      if (blockInfo.getBlockReader() != null) {
+        return blockInfo.getBlockReader();
+      }
+    } finally {
+      mLock.unlock();
+    }
+    BlockReader reader =
+        UnderFileSystemBlockReader.create(blockInfo.getMeta(), offset, noCache, mAlluxioBlockStore);
+    blockInfo.setBlockReader(reader);
```

Comment:
Does this portion need to be in the lock as well? What prevents the reader from being created multiple times?